### DISCUSSION TITLE
Duplicate flow inpossible?

### DISCUSSION THREAD

**Discussion by masterwishx** [goauthentik/authentik#9422] [Q&A] [answered]
I need to make another flow same as I have but with a little changes.
Is any way I can duplicate the flow or only create new one?

**Comment by BeryJu** [comment]
You can export a flow as a blueprint which is just YAML, and then modify it (currently you'll have to manually adjust/remove all the primary keys) and then re-import it (see https://docs.goauthentik.io/developer-docs/blueprints/ and https://github.com/goauthentik/authentik/blob/main/blueprints/default/flow-default-authentication-flow.yaml)

**Reply by masterwishx** [reply]
As i tryed to export flow , changed names then import to flow it just replace current because of same pk: .
if tryed make new flow then change to this pk: in imported file then it has 0 Stages ...

in blueprint if i want import exported flow should i change manual all `pk:` to `!KeyOf` ?

**Reply by masterwishx** [reply]
Also i cant edit default context in blueprint?  

![image](https://github.com/goauthentik/authentik/assets/28630321/bacc7068-ebc9-47bc-9bc5-a1a85fc6bb78)

**Reply by masterwishx** [reply]
Also tryed to export all flows to blueprints by `docker exec -i authentik-worker ak export_blueprint` but have error :  https://github.com/goauthentik/authentik/issues/8684

**Comment by kit-qt** [comment]
Did you make any progress with this? I have tried updating primary keys and slugs but end up with one flow having 0 stages or otherwise not being valid. All I'm after is the ability to have different enrolment flows for a few different groups of users.

**Reply by masterwishx** [reply]
> Did you make any progress with this? I have tried updating primary keys and slugs but end up with one flow having 0 stages or otherwise not being valid. All I'm after is the ability to have different enrolment flows for a few different groups of users.

No

**Comment by ravensorb** [comment]
Any chance of a "copy" feature being added?

**Comment by ching-kuo** [comment]
+1 on this. Sometimes I only want to modify part of the flow for a special scenario. It shouldn't be needed to recreate the whole flow itself.

**Comment by AzimovIz** [comment]
+1

**Comment by michaelkuty** [comment]
+1

**Comment by Pzdrs** [comment]
+1

**Comment by mihaly044** [comment]
+1

**Comment by chris4prez** [comment]
the ability to copy / duplicate flows within the UI missing seems to be such a basic necessity for ability to test and revert without needing to spend time performing manual manipulations via export/imports.

**Comment by shaunakv1** [comment]
+1

**Comment by RayBishopTN** [comment]
+1 This would be a great feature for those who typically don't use the default.

**Comment by Profecy** [comment]
+1 need this

**Comment by HubEight** [comment]
+1

**Comment by koresent** [comment]
+1

**Comment by rohrschacht** [comment]
+1

**Comment by realgeneraliroh** [comment]
+1

**Comment by binder-badge** [comment]
+1 would also love this

**Comment by JustHomelabbing** [comment]
+1

**Comment by goinsandrew** [comment]
+1. This is also great to help minimize n00bs like myself editing a default stage and destroying everything. The ability to duplicate and edit flows, stages and prompts would make life vastly simpler. 

For more advanced users, it would let you keep the entire setup you built, but only tweak what it needed (IE build your prompts, stages, and the initial flow for a default, then duplicate the flow, and tweak it for Site A, Site B, Site C, etc.

**Comment by ravensorb** [comment]
Its interesting -- its be 2 years and no maintainer has even answered this one with there own perspective

**Comment by dewi-tik** [comment, marked as answer]
We see how this would be useful and want to add it, but it's much more involved and complicated than you'd expect. A flow isn’t a single object, it’s a flow plus stage bindings, shared stages, and policies all wired together. Cloning that safely means deciding which pieces to duplicate, which to keep shared, and making sure that editing the copy doesn’t accidentally change the original. Doing that reliably needs a careful cloning mechanism and updated UI.

At the moment you can already somewhat achieve this by exporting a flow, editing the blueprint file, and then importing it again. This obviously isn’t streamlined/easy. But we’re working on improving the blueprint system so that it can eventually be used for this ‘duplicate flow’ mechanism.

We’re already tracking this change internally but we’ve created this [feature request](https://github.com/goauthentik/authentik/issues/23797) so that it can be tracked publicly, rather than in this discussion.

**Comment by ravensorb** [comment]
@dewi-tik - thank you for the update! This is great news.
